Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.cpp

```diff
@@ -0,0 +1,325 @@
+    {
+        m_beltStopped = true;
+    }
+
+    bool ConveyorBeltComponent::IsBeltStopped()
+    {
+        return m_beltStopped;
+    }
+
+    AZStd::pair<AZ::Vector3, AZ::Vector3> ConveyorBeltComponent::GetStartAndEndPointOfBelt(AZ::ConstSplinePtr splinePtr)
+    {
+        AZ::Transform splineTransform = AZ::Transform::Identity();
+        AZ::TransformBus::EventResult(splineTransform, m_entity->GetId(), &AZ::TransformBus::Events::GetWorldTM);
+        const AZ::SplineAddress addressBegin = splinePtr->GetAddressByFraction(0.f);
+        const AZ::SplineAddress addressEnd = splinePtr->GetAddressByFraction(1.f);
+        const AZ::Vector3 posBegin = splinePtr->GetPosition(addressBegin);
+        const AZ::Vector3 posEnd = splinePtr->GetPosition(addressEnd);
+        return { m_splineTransform.TransformPoint(posBegin), m_splineTransform.TransformPoint(posEnd) };
+    }
+
+    AZ::Transform ConveyorBeltComponent::GetTransformFromSpline(AZ::ConstSplinePtr splinePtr, float distanceNormalized)
+    {
+        AZ_Assert(splinePtr, "Spline pointer is null");
+        AZ::SplineAddress address = splinePtr->GetAddressByFraction(distanceNormalized);
+        const AZ::Vector3& p = splinePtr->GetPosition(address);
+
+        // construct the rotation matrix from three orthogonal vectors.
+        const AZ::Vector3& v1 = splinePtr->GetTangent(address);
+        const AZ::Vector3& v2 = splinePtr->GetNormal(address);
+        const AZ::Vector3 v3 = v1.Cross(v2);
+
+        AZ::Matrix3x3 rotationMatrix = AZ::Matrix3x3::CreateFromColumns(v1, v2, v3);
+        AZ_Assert(rotationMatrix.IsOrthogonal(0.001f), "Rotation matrix is not orthogonal");
+        rotationMatrix.Orthogonalize();
+        AZ::Transform transform = AZ::Transform::CreateFromMatrix3x3AndTranslation(
+            rotationMatrix, p - AZ::Vector3::CreateAxisZ(m_configuration.m_segmentSize / 2.f));
+        return m_splineTransform * transform;
+    }
+
+    float ConveyorBeltComponent::GetSplineLength(AZ::ConstSplinePtr splinePtr)
+    {
+        AZ_Assert(splinePtr, "Spline pointer is null");
+        AZ::Transform splineTransform = AZ::Transform::Identity();
+        AZ::TransformBus::EventResult(splineTransform, m_entity->GetId(), &AZ::TransformBus::Events::GetWorldTM);
+        const AZ::SplineAddress addressEnd = splinePtr->GetAddressByFraction(1.f);
+        return splinePtr->GetLength(addressEnd);
+    }
+
+    void ConveyorBeltComponent::MoveSegmentsGraphically(float deltaTime)
+    {
+        if (m_beltStopped)
+        { // Do not move texture when stopped
+            return;
+        }
+
+        // Animate texture
+        AZ::Render::MaterialAssignmentId materialId;
+        AZ::Render::MaterialComponentRequestBus::EventResult(
+            materialId,
+            m_configuration.m_conveyorEntityId,
+            &AZ::Render::MaterialComponentRequestBus::Events::FindMaterialAssignmentId,
+            -1,
+            m_configuration.m_graphicalMaterialSlot);
+
```

Comment:
Should we validate if a materialId was found here? (like Assert?)

Reply:
I did not add Assert but decided to add a check before that validate configuration.